llext: add support for SLID-based linking

mathieuchopstm opened the pull request:
This PR introduces support for an alternate linking method in the LLEXT subsystem, called "SLID" (short for *Symbol Link Identifier*), enabled by the **`CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID`** Kconfig option.

SLID-based linking uses a unique identifier (called "SLID") to identify exported symbols, instead of using the symbol name as done currently. In this PR, SLIDs are the first pointer-sized bits of `SHA256(symbol name)` and collisions are treated as a fatal error, but this is an implementation detail and could be changed (barring the ABI breakage this would cause).

This approach provides several benefits:
 * linking is faster because the comparison operation to determine whether we found the correct symbol in the export table is now an integer comparison, instead of a string comparison
    * N.B.: this PR does *not* add code to lookup symbols using a binary search algorithm, which would provide another significant performance boost (`O(log2 n)` comparisons vs `O(n)` currently).
    * N.B. 2: Binary search *could* be implemented without SLID linking by sorting the export table alphabetically, however, this requires ensuring that the sort algorithm is *exactly identical* to what `strcmp` performs at run-time. For this reason, export table is currently left unmodified when `CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID` is not enabled.
 * Zephyr binary size is reduced as symbol names can be dropped from the binary
 * confidentiality is improved as a side-effect, as symbol names are no longer present in the binary
   * N.B.: symbol names are ***not*** stripped from ELF by anything in this PR. I am unsure if this can be done automatically by using some ELF/object manipulation tool, or if manual ELF modification would be required.

This PR does not affect linking between different LLEXTs, only between LLEXTs and the Zephyr application. Cross-LLEXT linking via SLID should also be possible though.

This PR also reserves bit `0x08000000` of the `Elf_Shdr` `st_flags` field, as `SHF_ZEPHYR_RTOS_LLEXT_PREPARATION_DONE`. This is used to prevent parsing the same ELF twice, as the preparation process is destructive in the case of a Zephyr ELF if `CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID` is enabled. I have not found any usage of operating system-specific bits of the ELF format, but maybe there are already users and this bit is conflicting. Either way, a documentation note should probably be introduced somewhere, but I don't know exactly where it would be the most relevant.

Only implemented for ARM now, as I couldn't seem to find a single linker script or snippet that would work for all architectures, but I'm unsure whether there is no way to do so or I just haven't found how.

mathieuchopstm committed:
llext: add support for SLID-based linking

This commit introduces support for an alternate linking method in the
LLEXT subsystem, called "SLID" (short for Symbol Link Identifier),
enabled by the CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID Kconfig option.

SLID-based linking uses a unique identifier (integer) to identify
exported symbols, instead of using the symbol name as done currently.
This approach provides several benefits:
 * linking is faster because the comparison operation to determine
   whether we found the correct symbol in the export table is now an
   integer compare, instead of a string compare
 * binary size is reduced as symbol names can be dropped from the binary
 * confidentiality is improved as a side-effect, as symbol names are no
   longer present in the binary

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5412,8 +5412,29 @@ function(add_llext_target target_name)
-  set(noop_cmd ${CMAKE_COMMAND} -E true)
-  set(copy_cmd ${CMAKE_COMMAND} -E copy ${llext_lib_output} ${llext_pkg_input})
+
+  # LLEXT ELF processing for SLID imports (if enabled)
+  # Add a "hook" in the post-build commands
+  if (CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID)
+    set(noop_cmd
+      ${PYTHON_EXECUTABLE}
+      ${ZEPHYR_BASE}/scripts/build/inject_slids_in_llext.py
+      --elf-file ${llext_lib_output}
+      -vvv
+    )
+
+    set(copy_cmd
+      ${PYTHON_EXECUTABLE}
+      ${ZEPHYR_BASE}/scripts/build/inject_slids_in_llext.py
+      --elf-file ${llext_lib_output}
+      --output-file ${llext_pkg_input}
+      -vvv
+    )
+  else()
+    set(noop_cmd ${CMAKE_COMMAND} -E true)
+    set(copy_cmd ${CMAKE_COMMAND} -E copy ${llext_lib_output} ${llext_pkg_input})
+  endif()
```
I don't like this as it mixes up code related to `add_llext_command` processing with SLID, executing it before any `POST_BUILD` commands run. 
I think SLID could be considered "packaging", as it does mangle the ELF names so any processing on it becomes much harder. What do you think about moving to the packaging step? 
It could be something like adding an optional `slid_cmd`:
```
  if (CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID)
    set(slid_cmd
      ... --elf-file ${llext_pkg_input} ...
    )
  else()
    set(slid_cmd ${CMAKE_COMMAND} -E true)
  endif()
```

and then later adding 
```
   COMMAND $(slid_cmd)
```
as first step to all packaging steps?

mathieuchopstm replied:
I did think about this but couldn't think of a clean way to do it. This looks good to me though so I'll modify.

Nit: currently, `inject_slids_in_llext.py` is idempotent as it does *not* strip strings from the binary.
I did add [a check that prevents running it twice on the same LLEXT](https://github.com/mathieuchopstm/zephyr/blob/4331575936b5a8240266f69decb23ffdcf2e9e8a/scripts/build/inject_slids_in_llext.py#L100-L104), but it's an arbitrary limitation (safeguard in case something actually uses `st_value`, but AFAIK, the field is always unused for imported symbols).

pillo79 reviewed (commented):
Thanks for this! A few nits and discussion ideas.

mathieuchopstm commented on include/zephyr/llext/symbol.h:
```diff
@@ -27,14 +28,27 @@ extern "C" {
 struct llext_const_symbol {
-	/** Name of symbol */
-	const char *const name;
+	/** At build type, we always write to 'name'.
```
Fixed.

mathieuchopstm commented on tests/subsys/llext/simple/src/test_llext_simple.c:
```diff
@@ -24,6 +24,20 @@ LOG_MODULE_REGISTER(test_llext_simple);
+#define z_impl_ext_syscall_fail_SLID ((const char *)0xD58BC0E7ull)
+#endif
+
+#ifdef CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID
```
Indeed. Fixed.

(I did think about automatically generating these at build-time, but it would cause issues on fresh builds, because you would need the definitions to be able to build to be able to generate the definitions...)

teburd commented:
I love the idea, this is a great addition. If you have trouble with the linters I'd recommend reaching out on discord to hopefully get some clarifications. I agree the linter message in this case is not very clear.

pillo79 commented on cmake/modules/extensions.cmake:
```diff
@@ -5412,21 +5412,37 @@ function(add_llext_target target_name)
   # expected `pkg_input` location. If actual POST_BUILD commands are defined,
   # they will take care of this and the default copy is replaced by a no-op.
   set(has_post_build_cmds "$<TARGET_PROPERTY:${llext_proc_target},has_post_build_cmds>")
+
   set(noop_cmd ${CMAKE_COMMAND} -E true)
   set(copy_cmd ${CMAKE_COMMAND} -E copy ${llext_lib_output} ${llext_pkg_input})
+
```
Whitespace leftover from previous iteration?

mathieuchopstm replied:
Indeed. Removing it.

pillo79 commented on subsys/llext/Kconfig:
```diff
@@ -65,6 +65,25 @@ config LLEXT_STORAGE_WRITABLE
+	  up symbols from the built-in table by name. It also
+	  requires the LLEXTs to be post-processed after build.
+
+	  This option does not affect the behavior of linking
+	  symbols between multiple LLEXTs, only between LLEXTs
+	  and the Zephyr kernel / application.
```
Linking between llexts is not supported at the moment, but it makes sense to keep this for the future if it's valid. However, it's not clear to me why linking between llexts would be any different than with the main Zephyr binary. Can you explain a bit more?

mathieuchopstm replied:
I assumed that linking between LLEXTs had been implemented, for some reason. Mistake on my part.

As for why linking is different:
* linking to Zephyr is done by parsing the built-in export table, which is just a section filled with `struct llext_const_symbol`s
* linking to an LLEXT is done by parsing the ELF symbol table

There is however no huge technical difficulty to implement SLID linking for cross-LLEXT linking, but it is just not implemented by this PR so I wanted to make that clear (hence the option name, too).

Should this message be removed / reworded?

pillo79 replied:
Thanks for the explanation. Yes, there is currently no inter-llext linking and I'm not sure using the symbol table then is a good choice (I think it would easily cause mis-linking issue; I would vote for restricting to symbols in the export table). I would say this should be removed ATM.

mathieuchopstm replied:
You are right, I had forgotten that LLEXTs also have an export table, which is indeed what is used by `llext_find_sym()` when looking up symbols:
https://github.com/zephyrproject-rtos/zephyr/blob/149df6b61b6b3baee686cd62c9eb92237ff58d84/subsys/llext/llext.c#L127-L132

where the `sym_table` parameter comes from `struct llext`'s `exp_tab` field, populated here from `.exported_sym` section:
https://github.com/zephyrproject-rtos/zephyr/blob/149df6b61b6b3baee686cd62c9eb92237ff58d84/subsys/llext/llext.c#L502-L531
(Nit: this code appears to create a copy of the export table, which sounds like a huge RAM waste? But that's out of this PR's scope)

So this part of my comment was wrong, but I don't think it invalidates the point I was trying to make, i.e., the built-in export table is treated differently from LLEXTs', hence allowing to implement SLID-based linking for one or the other independently.


I'll remove the message.

pillo79 commented on include/zephyr/linker/llext-sections.ld:
```diff
@@ -0,0 +1,19 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+	/*
+	 * Special section used by LLEXT if CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID
+	 * is enabled. Declare this section to prevent it from being considered orphan.
+	 *
+	 * This section is used to temporarily save the exported symbols' names in the
+	 * Zephyr ELF for post-processing, but it is not included in the final binary.
+	 *
+	 * NOTE: This section MUST start at address 0, as the post-processing scripts
+	 * assume that the address of any data in this section (i.e., symbol names) is
+	 * strictly equivalent to the offset inside the section.
+	 */
```
Maybe this should not be indented?

mathieuchopstm replied:
I indented it to follow what appears to be a convention, such as in
https://github.com/zephyrproject-rtos/zephyr/blob/d77dc62d196b8f5f58622a05414002cbf40285ba/include/zephyr/linker/rel-sections.ld#L1-L35 or https://github.com/zephyrproject-rtos/zephyr/blob/d77dc62d196b8f5f58622a05414002cbf40285ba/include/zephyr/linker/kobject-text.ld#L1-L15

pillo79 reviewed (commented):
Thanks, I like the way this is going! A few additional debatable nit-pickings, and one last question I have on the "linking between llexts" thing.

mathieuchopstm commented on scripts/build/prepare_llext_exptab.py:
```diff
@@ -0,0 +1,374 @@
+        # might be possible and desireable.
+        #
+        # As of writing, this function will never be called as this script
+        # is only called if
```
Woops! Will add the missing part.

mathieuchopstm commented:
> Leaving additional commits unsquashed for now, will squash all before merge.
> 
> No idea how to pass the following CI check
> 
> ```
>  R0801:Similar lines in 2 files
> File:scripts/build/prepare_llext_exptab.py
> Line:1
> Column:0
> ```
> 
> Unless I misunderstand, it looks like it's complaining about the shebang, but that is _expected_ to be identical in all python scripts?
> 
> Edit: or the two docstrings are too similar?

Docstrings updated in 48d11a80bb815b895f0a48964e222ad38edd2986 didn't help, and I do not see how they can be considered "similar", so it most definitely looks like shebang is rejected. Asked for help on Discord (`#pr-help` channel).

pillo79 commented:
I managed to investigate this by manually calling `pylint -v -v -v scripts/build/llext_*` (not using the CI's `pylintrc`). This output suggests other lines may be triggering the failure:

```
scripts/build/llext_slidlib.py:1:0: R0801: Similar lines in 2 files
==llext_inject_slids:[135:165]
==llext_prepare_exptab:[333:363]
    parser.add_argument("-v", "--verbose", action="count",
                        help=("enable verbose output, can be used multiple times "
                              "to increase verbosity level"))
    parser.add_argument("--always-succeed", action="store_true",
                        help="always exit with a return code of 0, used for testing")

    return parser.parse_args(argv)

def _init_log(verbose):
    """Initialize a logger object."""
    log = logging.getLogger(__file__)

    console = logging.StreamHandler()
    console.setFormatter(logging.Formatter("%(levelname)s: %(message)s"))
    log.addHandler(console)

    if verbose and verbose > 1:
        log.setLevel(logging.DEBUG)
    elif verbose and verbose > 0:
        log.setLevel(logging.INFO)
    else:
        log.setLevel(logging.WARNING)

    return log

def main(argv=None):
    args = _parse_args(argv)

    log = _init_log(args.verbose)
 (duplicate-code)
```
```
scripts/build/llext_slidlib.py:1:0: R0801: Similar lines in 2 files
==llext_inject_slids:[117:128]
==llext_prepare_exptab:[317:328]
        res = self._prepare_inner()
        self.elf_fd.close()
        return res

def _parse_args(argv):
    """Parse the command line arguments."""
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
        allow_abbrev=False)
 (duplicate-code)
```
(no idea why the pylint error references slidlib but actually refers to other files :man_shrugging:)

edersondisouza commented on scripts/build/prepare_llext_exptab.py:
```diff
@@ -0,0 +1,365 @@
+
+        #1) Load the export table
+        exports_list = []
+        for (name_ptr, export_address) in self.exptab_manipulator: #TODO: this works?
```
Slightly concerning TODO here.... =D

mathieuchopstm replied:
Indeed :D
Leftover comment, script seems to work fine, will be removing it.

edersondisouza reviewed (commented):
Looks quite nice!

I'm curious as why the fixup commits are not squashed to the original ones? I think having fixup commits in the PR is not Zephyr way of doing this.

mathieuchopstm commented on cmake/modules/extensions.cmake:
```diff
@@ -5412,8 +5412,29 @@ function(add_llext_target target_name)
+  # LLEXT ELF processing for SLID imports (if enabled)
+  # Add a "hook" in the post-build commands
+  if (CONFIG_LLEXT_EXPORT_BUILTINS_BY_SLID)
+    set(noop_cmd
```
This was modified following pillo79's suggestion in e34ed801ec3c77daeca85c04c1f3a80014344a67.

I think you may be reviewing only the first commit and probably need to switch to "All commits" in diff view 
![diff view commit range selector](https://github.com/zephyrproject-rtos/zephyr/assets/147814239/e4d664dd-6c14-434b-a4a4-7fca9144e10d)

mathieuchopstm commented:
> I'm curious as why the fixup commits are not squashed to the original ones? I think having fixup commits in the PR is not Zephyr way of doing this.

As I mentioned a few comments earlier:
> Leaving additional commits unsquashed for now, will squash all before merge.

Please let me know if this makes review harder, I'll squash if so.

teburd commented:
> > I'm curious as why the fixup commits are not squashed to the original ones? I think having fixup commits in the PR is not Zephyr way of doing this.
> 
> As I mentioned a few comments earlier:
> 
> > Leaving additional commits unsquashed for now, will squash all before merge.
> 
> Please let me know if this makes review harder, I'll squash if so.

People do review commits, generally you’ll get a lot more commits once this is moved from draft and ci passes. At that point the commits should be the ones you want merged.

mathieuchopstm commented:
> People do review commits, generally you’ll get a lot more comments/reviews once this is moved from draft and ci passes. At that point the commits should be the ones you want merged.

I was planning to indeed squash when moving from draft to open PR. Didn't realize that people could be doing per-commit review, for which squashing indeed makes everything simpler.

Changelog:
* Squashed all commits
* pylint CI failure worked around, by adding `# pylint disable` comments at proper places 
  * linting locally with `pylint --rcfile=scripts/ci/pylintrc scripts/build/llext_*` no longer errors out at least
* Twister CI failures should be fixed
  * I had missed many linker scripts when adding the section, including Xtensa, which seems to be the cause of the issue. Hopefully I didn't forget any other...

pillo79 commented:
@mathieuchopstm This really looks promising. The issues identified by CI were due to the Apollo boards missing some include files, as mentioned in the above linked issue we can ignore the platform in the `testcase.yaml` for the time being until they provide them. Can you please update the PR?

mathieuchopstm committed:
tests: subsys: llext: exclude apollo4 platform

Exclude the Apollo4 platform from LLEXT tests for now,
as they currently break CI due to #72775.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
Hi @pillo79,
I think it should be done with the newly pushed commit. 

However, the `modules_enabled_writable_relocatable_slid_linking` test appears to fail on Xtensa, see:
* https://github.com/zephyrproject-rtos/zephyr/actions/runs/9079517508/job/24949489169?pr=72358#step:12:262
* https://github.com/zephyrproject-rtos/zephyr/actions/runs/9079517508/job/24949489169?pr=72358#step:12:1302

Looking at the logs, there is a suspicious line `W: PLT: cannot find idx 12 name printk`.
This is probably related to Xtensa using the *shared library* format for LLEXTs. I haven't had time to take a closer look yet; I'm not sure I'll be able to dig deep into this though, as I am not familiar with the Xtensa arch nor debugging of Zephyr under QEMU (and I don't have Xtensa hardware on hand).

pillo79 commented:
I have tried running it in Twister locally and figured out the issues:
* the various ELF mangling tools such as `elfconvert` and `strip` drop the SLID information, so the `inject_slid` _has_ to be the very last command or the `sym.st_value` in dynamic files is always 0; 
* there's an additional place where the SLID `#ifdef` had to be added due to the code duplication for the dynamic path :slightly_frowning_face: 
With these 2 changes my local twister does not report any issue :partying_face:  

While reviewing I also found a possible copy-paste error in `boards/qemu/x86/qemu_x86_tiny.ld` - there's no mention of llext and a duplicate line. We don't support x86 yet but it makes sense to keep this tidy :slightly_smiling_face:  

I mistakenly pushed my commits to your PR branch - they are obviously WIP so please feel free to squash them (or do your own implementation.

mathieuchopstm commented:
Very good job on the investigation @pillo79 ! And good catch for the x86 linker script error.

I have integrated the content of your commits and squashed them, following your suggestion of a single macro selecting the correct (name / SLID) based on Kconfig.

Builds and runs on my side, Twister on `Nucleo-F401RE` & `qemu_xtensa` both pass all non-skipped tests. `checkpatch` reports no obvious style problems either... let's cross fingers for all green CI :crossed_fingers:

dapperlo commented:
@mathieuchopstm  Very cool, and I agree it looks promising 

@mathieuchopstm @edersondisouza  What do we think about including the generated `slid_listing.txt` in the recently added EDK package?
That way the expected symbol translations are available to "standalone" extension developers for debug purposes. It's not strictly necessary since the translation is deterministic but I can see it being useful shortcut debug and something I'd expect to be delivered in a collateral package.

mathieuchopstm commented:
> What do we think about including the generated `slid_listing.txt` in the recently added EDK package? That way the expected symbol translations are available to "standalone" extension developers for debug purposes. It's not strictly necessary since the translation is deterministic but I can see it being useful shortcut debug and something I'd expect to be delivered in a collateral package.

I have not looked in details at how the EDK works; in fact, it is not supported by this PR's content at the moment.
I am assuming (and please correct me if this is wrong) that the `llext-edk` target requires a Zephyr application to be built first. In this case, wouldn't it be trivial to take the generated `slid_listing.txt` and bundle it in the EDK?

Otherwise, I assume that building SLID-linked LLEXTs with the EDK would imply that `llext_slidgen.py` and `llext_inject_slids.py` are included in it; this would make it easy to generate a SLID listing during the LLEXT build process. (One downside to this approach, though, is that exported-but-not-imported symbols would not have their SLID listed)

pillo79 reviewed (commented):
CI is finally green! :+1: 
@mathieuchopstm would you like to make this PR ready for review/merge?
(also, I had a pending comment for a very tiny nit)

mathieuchopstm commented:
> would you like to make this PR ready for review/merge?

LGTM, opening for review.

I have left all EDK-related questions unhandled for now after trying it on my side; since it only ships headers and CFLAGS, integrating SLID injection does sound non-trivial (as script depends on quite a few libraries provided by Zephyr, and ideally would be integrated in the build system). This could be worked on in a follow-up PR.

WRT EDK, I did think about failing the build if attempting to build it when SLID Kconfig option is enabled, but I don't really know what would be the "correct" way of doing this (replace the target's command with `cmake -E false`?)

pillo79 commented:
> I have left all EDK-related questions unhandled for now after trying it on my side; [..] This could be worked on in a follow-up PR.

Fully agree with this.

> WRT EDK, I did think about failing the build if attempting to build it when SLID Kconfig option is enabled, but I don't really know what would be the "correct" way of doing this (replace the target's command with `cmake -E false`?)

I think a better solution would be to stop the EDK generation from running if SLID is configured via an `if (SLID) message(FATAL_ERROR ...)` added in `cmake/llext-edk.cmake` .

mathieuchopstm commented:
CI all green except Documentation Build, but that seems unrelated to the PR's content (runner network issue?).

pillo79 commented on CMakeLists.txt:
```diff
@@ -1888,6 +1888,21 @@ if(CONFIG_BUILD_OUTPUT_INFO_HEADER)
+    --slid-listing ${PROJECT_BINARY_DIR}/slid_listing.txt
+    -vvv
+  )
+
```
Nit: remove empty line

pillo79 reviewed (approved):
Thanks for this contribution! Approving as-is; if further changes are requested, please fix a very tiny nit.

teburd reviewed (approved):
Fantastic change, thank you for contributing
